### ISSUE TITLE
[HOLD for payment 2023-01-11] IOS - Workspace -Units under Reimbursement-hidden by keyboard after tap on it (small phones)

### ISSUE THREAD

**Issue by kbecciv** [Expensify/App#11391]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Launch the app
2. Log in with any account
3. Go to Setting - Workspace
4. Tap on Reimbursable expenses
5. Tap on Units to change it

## Expected Result:
Units under Reimbursement should be visible after tap on it

## Actual Result:
Units under Reimbursement-hidden by keyboard after tap on it

## Workaround:
Unknow

## Platform:

Where is this issue occurring?

- iOS

**Version Number:** 1.2.9.0

**Reproducible in staging?:** Yes

**Reproducible in production?:** Yes

**Email or phone of affected tester (no customers):** any

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/93399543/192888345-9a5d8c1e-91cd-4e3a-b548-4348b3dac811.MP4

**Expensify/Expensify Issue URL:** 

**Issue reported by:** Applause - Internal Team

**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by JmillsExpensify** [comment]
Putting a hold on this and filing with our larger keyboard initiative.

**Comment by JmillsExpensify** [comment]
Moved to weekly while issue is on hold pending larger keyboard updates.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by AndrewGable** [comment]
Ok two weeks of not being able to reproduce means I think we should close, reopen if you disagree!

**Comment by kbecciv** [comment]
@JmillsExpensify @AndrewGable QA team is discovered the same issue in Android on build 1.2.17.0

https://user-images.githubusercontent.com/93399543/196445157-1d5f16b6-780d-483d-982b-c45e041fd044.mp4

**Comment by JmillsExpensify** [comment]
Still on hold for the larger keyboard issue mentioned in the title.

**Comment by JmillsExpensify** [comment]
@tgolen Let's take this one off hold right?

**Comment by tgolen** [comment]
It looks like the next step here is to get some proposals for the fix.

**Comment by JmillsExpensify** [comment]
Yes indeed! Upwork job is here: https://www.upwork.com/jobs/~01ff57d54ca13f4541.

**Comment by priyeshshah11** [comment]
**PROPOSAL:** 

This can be solved by using the [KeyboardAvoidingView](https://reactnative.dev/docs/keyboardavoidingview) with the 'behaviour' setting in `BlockingView` component https://github.com/Expensify/App/blob/main/src/components/BlockingViews/BlockingView.js#L29

**Comment by parasharrajat** [comment]
@priyeshshah11 This is an incomplete proposal. Please mention:
1. root cause of the issue. 
2. How your proposal fix the issue?

**Comment by priyeshshah11** [comment]
PROPOSAL:

**Cause:** This is the default behaviour of react-native to pull the keyboard up (over the actual view) 
**Solution:** This can be solved by using the [KeyboardAvoidingView](https://reactnative.dev/docs/keyboardavoidingview) with the behaviour prop in BlockingView component https://github.com/Expensify/App/blob/main/src/components/BlockingViews/BlockingView.js#L29

The problem will be fixed as KeyboardAvoidingView will push the input above in the actual view and the keyboard will appear just beneath it. Thus, the user will be able to see the input clearly while interacting with and having the keyboard open.

@parasharrajat My apologies!

**Comment by parasharrajat** [comment]
It still does not explain how this change solves the issue on `Reimbursable` page. Were you able to reproduce this issue? What is the relation of `BlockingView` with the issue?

**Comment by priyeshshah11** [comment]
> It still does not explain how this change solves the issue on `Reimbursable` page. Were you able to reproduce this issue? What is the relation of `BlockingView` with the issue?

This is the [`WorkspaceReimburseView`](https://github.com/Expensify/App/blob/3bba09a7898354378d5162ffd1fa0d5d66bf05b8/src/pages/workspace/reimburse/WorkspaceReimburseView.js) page which uses [`FullPageNotFoundView`](https://github.com/Expensify/App/blob/3bba09a7898354378d5162ffd1fa0d5d66bf05b8/src/pages/workspace/reimburse/WorkspaceReimburseView.js#L193) which in turn uses [`BlockingView`](https://github.com/Expensify/App/blob/156fd61d8351a80bb30e0f65a0ee50ac0a8f27e2/src/components/BlockingViews/FullPageNotFoundView.js#L59)

I hope this explains the relation

**Comment by parasharrajat** [comment]
Thanks. It makes sense. But we are already have `KeyboardAvoidingView` in `ScreenWrapper` https://github.com/Expensify/App/blob/3bba09a7898354378d5162ffd1fa0d5d66bf05b8/src/pages/workspace/WorkspacePageWithSections.js#L100

**Comment by priyeshshah11** [comment]
@parasharrajat any updates on the above?

**Comment by parasharrajat** [comment]
I just checked. This is working fine for me on Android. 

I am facing some issues with my Mac ATM. I will share an update as soon as possible.

**Comment by parasharrajat** [comment]
> however it is using the default props which is currently set to padding for the behaviour prop in ScreenWrapper which adds a paddingBottom to the main view and not the scrollView which is being used inside [WorkspacePageWithSections](https://github.com/Expensify/App/blob/3bba09a7898354378d5162ffd1fa0d5d66bf05b8/src/pages/workspace/WorkspacePageWithSections.js#L112). Thus, the content inside the scrollView stays as it is and doesn't move up, we need to scroll the content up as well inside the ScrollView.

Based on this, you are saying that KeyboardAvoindingView is not working. I don't think this is correct. It is working for the `Rate` field.  

--- 

Waiting for proposals.

**Comment by JmillsExpensify** [comment]
Still waiting for an approved proposal.

**Comment by JmillsExpensify** [comment]
Same as above. Increasing the price to $500.

**Comment by JmillsExpensify** [comment]
Ideally hoping for Margelo or an internal engineer to take this over. I'm holding on raising the price for the time being.

**Comment by tgolen** [comment]
I'm going to take this over and start digging into it

**Comment by tgolen** [comment]
I'm not able to reproduce this on either iOS or Android.

<img width="452" alt="2022-11-18_15-11-48" src="https://user-images.githubusercontent.com/1228807/202811823-f4b76b57-1730-4723-8289-0f783dc28cf2.png">
<img width="422" alt="2022-11-18_15-12-13" src="https://user-images.githubusercontent.com/1228807/202811834-963d1294-4e0e-49e0-8346-d969fe29fe07.png">

**Comment by parasharrajat** [comment]
@tgolen Did you test the Unit field? This issue focuses on that. Rate is used for comparison only.

**Comment by tgolen** [comment]
Maybe I'm confused on the reproduction steps. On Android, if I just go into the settings and tap on the units, I see this with no keyboard: 

<img width="422" alt="image" src="https://user-images.githubusercontent.com/1228807/202815879-644e9108-2ad8-46ce-96d3-9605bd76d418.png">

So... it's not really a valid test?

**Comment by parasharrajat** [comment]
This issue is only occurring at iOS.

**Comment by JmillsExpensify** [comment]
Oh ok. So we need to update the title/reproduction steps then for clarity.

**Comment by parasharrajat** [comment]
Yeah, sure. I think it was iOS some days back.

**Comment by parasharrajat** [comment]
Are we keeping it close? I saw you updated the title @JmillsExpensify. So I am unclear.

**Comment by tgolen** [comment]
Digging into this, I found that this is a bug in the picker library we are using. There is an open issue about it here: https://github.com/lawnstarter/react-native-picker-select/issues/273. The library looks to be unmaintained and hasn't had any real changes merged in several years. We also have our own fork of the library.

@hannojg This might be a good one for you to look into and investigate. Maybe you can find the cause of the issue in the library and we can fix it in our fork, as well as submit a PR to the original library to fix it also. It looks like there might be a workaround posted in https://github.com/lawnstarter/react-native-picker-select/issues/273 but I don't think we want to use that.

**Comment by tgolen** [comment]
Or maybe there is a better picker component we should be using?

**Comment by hannojg** [comment]
Oh yeah sure, will add that to my list! Will look into it this week. How urgent is this?

**Comment by tgolen** [comment]
I'd say it's on the low end of urgency. These pages aren't used very much and the impact of this bug is very little (ie. you can  still use the form perfectly fine).

**Comment by JmillsExpensify** [comment]
Agree with adding this as a weekly for now.

**Comment by JmillsExpensify** [comment]
Still a low priority, though on the list.

**Comment by JmillsExpensify** [comment]
Though this is still low on the priority list, @hannojg this is a good one to pick up this week so I'm going to assign you.

**Comment by JmillsExpensify** [comment]
He's out sick so expect him to action on this in the coming days.

**Comment by chrispader** [comment]
Here i am, GitHub. :-)

**Comment by JmillsExpensify** [comment]
Thank you! Assigned.

**Comment by chrispader** [comment]
# Proposal

## Problem

The problem with `KeyboardAvoidingView` and `react-native-picker-select` is, that the Picker is not perceived as a Keyboard in react native's `ScrollView` component.

`ScrollView` reacts to Keyboard open/close events and scrolls the content up/down by the minimum height thats needed to show the focused `TextInput` component.

The Picker internally is just a custom `Modal` that gets opened when you click on the `TextInput`. The workaround [here](https://github.com/lawnstarter/react-native-picker-select/issues/273#issuecomment-805998002) is trying to add this functionality, by adding a `onShowKeyboard` and `onCloseKeyboard` callback, measure the view itself and then use `scrollTo` on the corresponding `ScrollView` manually.

This is not necessarily a bug in the `react-native-picker-select` library, it's just a missing feature.

## Solution

Pass `ScrollView` ref to `Picker` component and scroll to input directly in `react-native-picker-select` component.

I have implemented such a behaviour here:
[react-native-picker-select](https://github.com/chrispader/expensify-react-native-picker-select/commit/5be620f56a032e9858827641e082081401fca0b4)
[Expensify/App](https://github.com/margelo/expensify-app-fork/commit/cc53825775e43fa32bf4a84fca0ef128f313bbcc)

**Comment by chrispader** [comment]
Result:

https://user-images.githubusercontent.com/20173411/205634114-ffc884f6-d21d-4d1c-b516-d8f37c084a16.mp4

**Comment by chrispader** [comment]
In this screen recording, the modal is opening kinda buggy the first time, but that's just the video

**Comment by JmillsExpensify** [comment]
@tgolen thoughts on @chrispader's findings and proposal?

**Comment by tgolen** [comment]
The findings make sense. Thanks for looking into it @chrispader! There are a few things that I don't really like about the solution, but I'm not sure if there is any choice if we want this fixed.

1. It feels brittle that we have to pass a scrollview ref to the picker. Someone can easily forget to do that, and then this isn't working. It looks like the ref has to be passed through a couple of layers too, which makes it even easier to miss.
3. I see the fix is really being implemented in our fork of `react-native-picker-select` and so I would also like to see a plan that includes getting it into the upstream and original repo, rather than us having to fight to maintain it forever in our fork

Is it maybe possible to implement the scrolling completely in `App` and in our own picker?

**Comment by chrispader** [comment]
Thanks for your input @tgolen !

1. We can either handle this behaviour over emitting events or by using the ref of the ScrollView. A way to prevent passing down the ref (through possibly multiple layers) would be use create a `ScrollViewContext` and use it in expensify's custom `Picker` component.

If we want to completely have this change in the `Expensify/App` repo, we would essentially be using the [workaround](https://github.com/lawnstarter/react-native-picker-select/issues/273#issuecomment-805998002) you mentioned earlier, but still we'd need a ref to the ScrollView.

2. I'll definitely set up a PR in the original repo, still gonna make some changes to the code

**Comment by tgolen** [comment]
> To the second point, i'll definitely set up a PR in the original repo, still gonna make some changes to the code though

👍 

> We could also wrap both RNSelectPicker and ScrollView with HOCs and handle all the logic around scrollTo in that HOC. Still, we'll need to add some sort of wrapper/reference to the ScrollView

Yeah, I think if we could hide the implementation details (eg. passing the scrollview ref) inside an HOC or a custom component, then I think that would be a little more ideal. 

Another consideration might be to use render-props and create a component like `<ScrollViewWithPickers>` that provides a `ref` that can be passed to the children. We could add some comments at the top of the `<ScrollViewWithPickers>` with instructions and an example that show how it should be used.

**Comment by JmillsExpensify** [comment]
Awesome convo! Are we squared away to start working on the PR?

**Comment by tgolen** [comment]
I'd still like to get a better idea of what the proposal would look like, but if @chrispader just wants to work in the PR to iron out the details, that's fine with me too.

**Comment by chrispader** [comment]
@JmillsExpensify @tgolen added a PR for this. in `Expensify/App`

The changes in `react-native-picker-select` i would like to submit to the general library, once it got reviewed here

**Comment by JmillsExpensify** [comment]
I wanted to give everyone a quick heads up that I'm OOO this week, though I intend to check in every couple of days. @tgolen if you need anything quickly on the Bug Zero side of things, it'd be great if you can handle it or raise it #bug-zero in Slack for help. Thanks!

**Comment by tgolen** [comment]
There will be three PRs for this issue:

1. [A PR](https://github.com/Expensify/react-native-picker-select/pull/3) to the Expensify fork of `react-native-picker-select` 
2. [A PR](https://github.com/Expensify/App/pull/13514) to App to implement the new features of the picker
3. A PR to the official `react-native-picker-select` upstream library

It sounds like `3` will be done after we have completed `1` and `2`, is that right @chrispader ?

**Comment by chrispader** [comment]
Yes, exactly! @tgolen

Working on both of these PRs (1 & 2) right now

**Comment by JmillsExpensify** [comment]
Circling back on 3, do we have a separate issue for that, or we're going to hold off closing this issue until all three PRs are merged (either in this repo or upstream)?

**Comment by JmillsExpensify** [comment]
Looks like we are still working through https://github.com/Expensify/App/pull/13514, which @parasharrajat was able to review today. Woo! That said, most everyone on this issue is out for the holidays, so I'm moving this issue to weekly and we can resume daily updates next week.

**Comment by chrispader** [comment]
I think this issue is already fixed, right? @tgolen @JmillsExpensify

**Comment by tgolen** [comment]
For the most part, yes. Did we ever do this?

> A PR to the official react-native-picker-select upstream library.

**Comment by chrispader** [comment]
I added a PR in the upstream library. You can find it [here](https://github.com/lawnstarter/react-native-picker-select/pull/496)

**Comment by JmillsExpensify** [comment]
Great point. So I think in that case, we need to: 

 - Add the BugZero checklist to this issue ✅ 
 - Create an issue to revert our rn-picker fork when the PR is merged upstream
 - So at this point we just need to close out the checklist and then issue payments.

**Comment by JmillsExpensify** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@parasharrajat / @tgolen] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@parasharrajat / @tgolen] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@parasharrajat / @tgolen] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@JmillsExpensify] A [regression test](https://expensify.testrail.io/index.php?/dashboard) has been added or updated so that the same bug will not reach production again. Link to the GH issue for creating the test here:

**Comment by JmillsExpensify** [comment]
Circling back to payments, this is where I think we landed. 

 - Issue reporter: N/A
 - Contributor: @chrispader $1,000
 - Contributor+: @parasharrajat $1,000

@parasharrajat the upwork job is here: https://www.upwork.com/jobs/~014479aca2e5c69cc7. @chrispader I think Margelo is adding these to the monthly invoice, so make sure you pass the word along.

**Comment by JmillsExpensify** [comment]
@parasharrajat @tgolen Do both of you mind giving us a head start on the BZ checklist above?

**Comment by tgolen** [comment]
Sure, this one is easy since it wasn't a regression. It was just something missing from the original implementation (so technically we added a feature).

**Comment by tayyabHussain7860** [comment]
you can solve the issue by wrapping the Textinput in KeyboardAvoidingView. 

`import { KeyboardAvoidingView } from 'react-native'
import { useHeaderHeight } from '@react-navigation/elements'

  const height = useHeaderHeight()

<KeyboardAvoidingView
      keyboardVerticalOffset={height + 47}
      behavior="padding"
      style={{ flex: 1 }}
      enabled>
     <TextInput
     placeholdertext="Enter Rate"

     />
    </KeyboardAvoidingView>

`

**Comment by tgolen** [comment]
@tayyabHussain7860 thanks for the proposal, but this has already been worked on and fixed. Please read the issue history closer in the future.

**Comment by JmillsExpensify** [comment]
@parasharrajat Let's close the loop on this one. I've gone ahead and invited you to the Upwork job.

**Comment by JmillsExpensify** [comment]
P.S. Regression issue is created, so we'll be ready to close this out as soon as Rajat is able to accept.

**Comment by JmillsExpensify** [comment]
@parasharrajat has been paid out, so we should be done here. Thanks!
